### PR TITLE
fix: escape arguments containing spaces

### PR DESCRIPTION
**Pull request by topliceanurazvan**
## Fixes
 
- Escape arguments in textarea with spaces

## How to test it

- Add arguments to a test that contain space characters

## screenshots

image

## Checklist

- [x] tested locally
- [ ] added new dependencies
- [ ] updated the docs
- [ ] added a test

### CODE REVIEW

File: packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx [hunk 1]
```diff
@@ -28,6 +28,14 @@ import {prettifyArguments} from '@utils/prettifyArguments';
 
 import {ArgumentsWrapper} from './Arguments.styled';
 
+const formatArgsArray = (args: string[]) => {
+  return args.map(arg => {
+    const trimmedArg = arg.replace(/\s+/g, ' ').trim();
```
Comment (rangoo94) [inline packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx:33]: It should rather not trim the argument, as the argument may start from spaces (rather edge case, but still):
```yaml
args:
- "   i want to have spaced argument  "
```

File: packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx [hunk 2]
```diff
@@ -28,6 +28,14 @@ import {prettifyArguments} from '@utils/prettifyArguments';
...
+
+    return arg.includes(' ') && !/^(['"]).*\1$/.test(arg) ? `"${trimmedArg}"` : trimmedArg;
```
Comment (rangoo94) [inline packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx:35]: Checking the starting character won't work, as:
```yaml
args:
- "some value"
```
Will result in:
```json
["some value"]
```

It will start from `['"]` only in case it specifically has it, and it should be persisted:
```yaml
args:
- "\"abc"
```

Additionally, the quotes inside needs to be escaped too. As an example `abra"kadabra` should not be escaped as `"abra"kadabra"`, but rather `"abra\"kadabra"`

File: packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx [hunk 3]
```diff
@@ -28,6 +28,19 @@ import {prettifyArguments} from '@utils/prettifyArguments';
 
 import {ArgumentsWrapper} from './Arguments.styled';
 
+const formatArgsArray = (args: string[]) => {
+  return args.map(arg => {
+    const trimmedArg = arg.replace(/\s+|(?<=[^'])['"](?=[^'])/g, match => {
+      if (match.trim() === '') {
+        return ' '; // Replaces multiple spaces with a single space
```
Comment (rangoo94) [inline packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx:35]: It should not replace multiple spaces in a context of single argument.

It would replace:
```yaml
# abc "   def  xyz   " "abc"
command:
  - abc
args:
  - "   def  xyz   "
  - "abc"
```

With:
```yaml
# abc " def xyz " "abc"
command:
  - abc
args:
  - " def xyz "
  - "abc"
```

File: packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx [hunk 4]
```diff
@@ -28,6 +28,19 @@ import {prettifyArguments} from '@utils/prettifyArguments';
...
+      }
+      return `\\${match}`; // Escapes quotes
```
Comment (rangoo94) [inline packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx:37]: It would work, although won't display nice. It will unnecessarily escape some characters.

As an example `"abc'def"` will become `"abc\'def"` despite, `\` is redundant here

File: packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx [hunk 5]
```diff
@@ -28,6 +28,10 @@ import {prettifyArguments} from '@utils/prettifyArguments';
 
 import {ArgumentsWrapper} from './Arguments.styled';
 
+const formatArgsArray = (args: string[]) => {
+  return args.map(arg => (arg.includes(' ') ? `"${arg}"` : arg));
```
Comment (rangoo94) [inline packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx:32]: Missing escaping of `"` inside.

```
args:
- 'abc" def'
```
will become invalid `"abc" def"` instead of `"abc\" def"`

File: packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx [hunk 6]
```diff
@@ -103,7 +105,7 @@ const Arguments: React.FC<ArgumentsProps> = ({readOnly}) => {
       onCancel={onCancel}
     >
       <ArgumentsWrapper>
-        <CopyCommand command={currentArgsInline} isBordered additionalPrefix="executor-binary" />
+        <CopyCommand command={currentArgs} isBordered additionalPrefix="executor-binary" />
```
Comment (rangoo94) [inline packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx:108]: Won't it make it multiline? Maybe sth like:

```
// Above
const prettifiedArgs = useMemo(() => prettifyArguments(currentArgs), [currentArgs]); // and replace `isPrettified` to not use `useMemo`
const currentArgsInline = useMemo(() => prettifiedArgs.replace(/\n+/g, ' '), [prettifiedArgs]);
```

File: packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx [hunk 7]
```diff
@@ -28,6 +28,10 @@ import {prettifyArguments} from '@utils/prettifyArguments';
 
 import {ArgumentsWrapper} from './Arguments.styled';
 
+const formatArgsArray = (args: string[]) => {
```
Comment (rangoo94) [inline packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx:31]: Could be worth to extract it as some `escapeArgument` function along with unit tests 👍

File: packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx (hunk 5 above)
Comment (rangoo94) [reply packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx:32]: Additionally, it sould escape quotes even when there is no space, otherwise:
```
args:
- 'abc"def'
```

will become `abc"def` instead of `abc\"def`

Comment (rangoo94) [reply packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx:32]: Plus, it should check for any kind of space `/\s/.test(arg)`

Comment (rangoo94) [reply packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx:32]: We could as well just use external library for quoting, like [**shell-quote**](https://www.npmjs.com/package/shell-quote)

File: packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx [hunk 8]
```diff
@@ -18,6 +18,8 @@ import {Permissions, usePermission} from '@permissions/base';
 
 import {useUpdateTestMutation} from '@services/tests';
 
+import {escapeArguments} from '@src/utils/escapeArguments';
```
Comment (rangoo94) [inline packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx:21]:
```suggestion
import {escapeArguments} from '@utils/escapeArguments';
```

File: packages/web/src/utils/escapeArguments.ts [hunk 9]
```diff
@@ -0,0 +1,7 @@
+export const escapeArguments = (args: string[]) => {
+  return args.map(arg => {
+    const escapedArg = arg.replace(/(")/g, '\\$1');
```
[the remark this answers is not in the document]
Comment (rangoo94) [reply packages/web/src/utils/escapeArguments.ts:3]: Still quoting of `'` missing (and not important: maybe backtick, new line), i.e. for `args: ["abc'def"]` it will output `abc'def` instead of `"abc'def"` or `abc\'def`

Maybe the simplest fix would be to change regex below to quote if `[\s']` is matching

File: packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx [hunk 10]
```diff
@@ -45,15 +46,18 @@ const Arguments: React.FC<ArgumentsProps> = ({readOnly}) => {
...
+  const initialArgs = useMemo(() => escapeArguments(entityArgs)?.join('\n') || '', [entityArgs]);
 
   const currentArgs = Form.useWatch('args', form) || '';
-  const currentArgsInline = currentArgs.replace(/\s+/g, ' ').trim();
-  const isPrettified = useMemo(() => currentArgs === prettifyArguments(currentArgs), [currentArgs]);
 
-  const onSaveForm = () => {
-    const values = form.getFieldsValue();
-    const argVal = values.args?.trim().split('\n').filter(Boolean) || [];
+  // Above
```
Comment (rangoo94) [inline packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx:53]:
```suggestion
```

File: packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx [hunk 11]
```diff
@@ -45,15 +46,18 @@ const Arguments: React.FC<ArgumentsProps> = ({readOnly}) => {
...
+  const prettifiedArgs = useMemo(() => prettifyArguments(currentArgs), [currentArgs]); // and replace `isPrettified` to not use `useMemo`
```
Comment (rangoo94) [inline packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx:54]:
```suggestion
  const prettifiedArgs = useMemo(() => prettifyArguments(currentArgs), [currentArgs]);
```

File: packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx [hunk 12]
```diff
@@ -45,15 +46,18 @@ const Arguments: React.FC<ArgumentsProps> = ({readOnly}) => {
...
+  const currentArgsInline = useMemo(() => prettifiedArgs.replace(/\n+/g, ' '), [prettifiedArgs]);
+
+  const isPrettified = Boolean(currentArgs === prettifyArguments(currentArgs));
```
Comment (rangoo94) [inline packages/web/src/components/pages/Tests/TestDetails/TestSettings/SettingsVariables/Arguments.tsx:57]: Unnecessary boolean cast, unnecessary `prettifyArguments` call:
```suggestion
  const isPrettified = currentArgs === prettifiedArgs;
```

File: packages/web/src/utils/escapeArguments.ts [hunk 13]
```diff
@@ -0,0 +1,7 @@
+export const escapeArguments = (args: string[]) => {
+  return args.map(arg => {
+    const escapedArg = arg.replace(/(["'`])/g, '\\$1');
+
+    return /\s/.test(escapedArg) ? `"${escapedArg}"` : escapedArg;
```
Comment (rangoo94) [inline packages/web/src/utils/escapeArguments.ts:5]: `abc' def` will result in `"abc\' def"` (unnecessary escaping of `'`)

```suggestion
    return /\s/.test(escapedArg) ? `"${escapedArg.replace(/\\'/g, '\''}"` : escapedArg;
```
+ unit test

**Review by rangoo94 — APPROVED** [review]
LGTM 👍
